### PR TITLE
API life cycle and deprecation policy in official documentation

### PR DESCRIPTION
**Pull request by Olivia-liu**
Summary: 

This diff adds a new page to the ET public documentation site that establishes the API life cycle and deprecation policy. It's also been linked to from related docs.
The doc has some TODO items we can either address in this diff or with follow-up diffs.

For Meta reviewers: to comment easily, please review in https://docs.google.com/document/d/1FJmsWoP544bF0Xe3SeREcLAbKPqgrxulScBdak0UoVs/edit?usp=sharing

Differential Revision: D60692061

### CODE REVIEW

File: docs/source/api-life-cycle.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+1. Discuss the change and collect initial feedback.
+2. Clearly mark the API deprecated in code and documentation (See “How to Mark?” below). The mark would trigger static and/or runtime warning that contains at least these information (or in documentation only, if a warning message is not feasible):
+    1. Clearly point to the non-deprecated alternative to migrate to, or be clear if there is no alternative.
+    2. Specify the earliest version in which the API may actually be removed (See “Deprecation Period” below)
+3. Watch out for feedback. Users who weren’t involved in the original discussion can now comment. Maybe reconsider the deprecation.
+4. The API removal may now be made permanent having reached the declared version. Make sure documentation is also cleaned up.
+
+
+In the case that a breaking change has to be made to an existing interface, for example, if a non-optional parameter has to be added to a method, then we treat this as a deprecation case. Steps to take:
```
Comment (dbort) [inline docs/source/api-life-cycle.md:91]: This sentence is a little hard to follow. Suggested rewording.

```suggestion
We also use deprecation as a way to make breaking changes to an existing interface: for example, if adding a non-optional parameter to a method. To do this without breaking existing users:
```

File: docs/source/api-life-cycle.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+
+1. Create a new API that meets the new requirements.
+2. Migrate use cases from the old API to the new API.
+3. Deprecate the old API.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:96]: Deprecation should happen at the same time as the introduction of the new API.

```suggestion
1. In a single commit:
  - Create a new API that meets the new requirements.
  - Deprecate the old API and recommend that users move to the new API.
2. Migrate use cases from the old API to the new API.
3. Delete the old API after the deprecation period.
```

File: docs/source/api-life-cycle.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+
+## How to Mark?
+
+We want to make the best effort to mark in the most standard way possible, so that it’s easier for users to integrate with their dev tools, like Linters.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:101]:
```suggestion
When possible, the ExecuTorch code uses language-standard ways to annotate API lifecycle state in the code. This makes it easier for IDEs and other tools to communicate state to developers.
```

File: docs/source/api-life-cycle.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+   <td><strong>Documentation</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>Python
+   </td>
+   <td>
+<ul>
+
+<li><em>Deprecated</em>: use <a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a> decorator
```
Comment (dbort) [inline docs/source/api-life-cycle.md:118]:
```suggestion
<li><em>Deprecated</em>: use the <a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a> decorator
```

File: docs/source/api-life-cycle.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+<li><em>Experimental</em>: use ExecuTorch's native experimental (TODO not yet implemented) decorator
+</li>
+</ul>
+   </td>
+   <td>
+<ul>
+
+<li>Use <code>.. warning::</code> in docstring of deprecated and experimental APIs, see example usage <a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">here</a>.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:127]:
```suggestion
<li>Use <code>.. warning::</code> in the docstrings of deprecated and experimental APIs. See <a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">example usage</a>.
```

File: docs/source/api-life-cycle.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+</ul>
+   </td>
+  </tr>
+  <tr>
+   <td>C++
+   </td>
+   <td>
+<ul>
+
+<li>Use the <code>__ET_DEPRECATED</code> and (TODO not yet implemented) <code>__ET_EXPERIMENTAL</code> macro, see example usage <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">here</a>.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:138]:
```suggestion
<li>Use the <code>__ET_DEPRECATED</code> or (TODO not yet implemented) <code>__ET_EXPERIMENTAL</code> macros. See <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">example usage</a>.
```

File: docs/source/api-life-cycle.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+</li>
+</ul>
+   </td>
+   <td>
+<ul>
+
+<li>Start docstring with “DEPRECATED” and “EXPERIMENTAL”, see example usage <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">here</a>.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:145]:
```suggestion
<li>Start Doxygen comments with "DEPRECATED:" or "EXPERIMENTAL:". See <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example usage</a>.
```

File: docs/source/api-life-cycle.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+### Enforcement
+
+We want to implement automated tests to monitor:
+* Were any BC-breaking changes made to stable APIs?
+* Did the removal of an deprecated API happen not sooner than the promised period?
+
+
+### Auto migration
+
+The current deprecation policy still requires users to take actions to migrate off from deprecated APIs. Ideally, we should automate this process. We want to leverage tools like [TorchFix](https://pypi.org/project/TorchFix/) to auto-migrate uses of deprecated APIs when possible.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:186]: Please remove this section. We do want to work on it, but it's not relevant to the policy itself.
```suggestion
```

**Review by dbort — CHANGES_REQUESTED** [review]
The structure looks great; thanks for adding this.

File: docs/source/api-life-cycle.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+   <td>
+
+Start Doxygen comments with "DEPRECATED:" or "EXPERIMENTAL:". See <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example usage</a>.
+
+   </td>
+  </tr>
+  <tr>
+   <td>Android
+   </td>
+   <td>Deprecated: use <a href="https://docs.oracle.com/javase/9/docs/api/java/lang/Deprecated.html">java.lang.Deprecated</a>
```
Comment (dbort) [inline docs/source/api-life-cycle.md:99]: Consistency: This uses non-italicized "Deprecated" and "Experimental", but the first row italicizes them. The C++ row doesn't even mention "Deprecated" or "Experimental" directly. And the Android/Apple rows only seem to talk about deprecated code, not experimental code.

Ideally, each row of this table would be consistent with all of the other rows.

File: docs/source/api-life-cycle.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+<p>
+<code>@available(*, deprecated, message: "Use newMethod instead")</code>
+   </td>
+   <td>In Objective-C:
+<p>
+/**
+<p>
+* @deprecated Use `newMethod` instead.
+<p>
+*/
```
Comment (dbort) [inline docs/source/api-life-cycle.md:133]: Would be nice to make these code blocks, though I don't know if that works in a table.
```suggestion
/**
<pre><code>
<p>
* @deprecated Use `newMethod` instead.
<p>
</code></pre>
*/
```

Or maybe triple-backquote will work here.

Same for the swift example

File: docs/source/api-life-cycle.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+1. Discuss the change and collect initial feedback.
+2. Clearly mark the API deprecated in code and documentation (See “How to Mark API State” below).
+3. Listen to user feedback after the first release that deprecates the API. Users who weren't involved in the original discussion may have good arguments for not deprecating or removing the API.
+4. Once the deprecation period has passed, the API may be removed (See “Deprecation Period” below). Be sure to also remove references from the documentation.
+
+
+We also use deprecation as a way to make breaking changes to an existing interface: for example, if adding a non-optional parameter to a method. To do this without breaking existing users:
+1. In a single commit:
+   - Create a new API that meets the new requirements.
```
Comment (dbort) [inline docs/source/api-life-cycle.md:47]: Consistency: the rest of the doc uses `*` for bullets, but this uses `-`

File: docs/source/api-life-cycle.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+Start Doxygen comments with "DEPRECATED:" or "EXPERIMENTAL:". See <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example usage</a>.
+
+   </td>
+  </tr>
+  <tr>
+   <td>Android
+   </td>
+   <td>Deprecated: use <a href="https://docs.oracle.com/javase/9/docs/api/java/lang/Deprecated.html">java.lang.Deprecated</a>
+<p>
+Experimental: use <a href="https://developer.android.com/reference/androidx/annotation/RequiresOptIn">androidx.annotation.RequiresOptIn</a>
```
Comment (dbort) [inline docs/source/api-life-cycle.md:101]: Maybe point to https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:docs/api_guidelines/annotations.md as well (or instead), which give more guidance about using this annotation.

File: docs/source/api-life-cycle.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+   </td>
+   <td>/**
+<p>
+@deprecated As of release x.x, replaced by {@link #newMethod()}
+<p>
+*/
+<p>
+/**
+<p>
+Warning: This API is experimental
```
Comment (dbort) [inline docs/source/api-life-cycle.md:111]: Our style guide prefers full sentences, ending with periods.
```suggestion
Warning: This API is experimental.
```

**Review by dbort — CHANGES_REQUESTED** [review]
Thanks for the changes! Remaining comments are about consistency throughout the doc.

File: docs/source/api-life-cycle.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+  </tr>
+  <tr>
+   <td>Objective-C
+   </td>
+   <td>
+<p>
+<code>__attribute__((deprecated("Use newMethod instead")));</code>
+<p>
+<p>
+<code>__attribute__((experimental("Use newMethod instead")));</code> (TODO not yet implemented)
```
Comment (dbort) [inline docs/source/api-life-cycle.md:173]: I don't know if it's possible to implement an attribute like this; they're typically defined by the toolchain.

**Review by dbort — APPROVED** [review]
Thanks for all of the changes. A couple final issues, but I'll approve now to move things along.

File: docs/source/api-life-cycle.md (hunk 14 above)
Comment (Olivia-liu) [reply docs/source/api-life-cycle.md:173]: I'll let @shoumikhin comment on this. I've chatted with him offline and IIUC, there's a mapping between the Objective-C annotations and the CPP macros. So if the CPP `__ET_DEPRECATED` and `__ET_EXPERIMENTAL` can be implemented, then we can implement these.

Comment (shoumikhin) [reply docs/source/api-life-cycle.md:167]: Oh I meant we can use the deprecated annotation for experental features too, since there's no dedicated experimental annotation available. Thus __ET_EXPERIMENTAL is gonna be defined the same as deprecated, but with a different error message.
